lib/cn/cn.c

```diff
@@ -773,28 +773,63 @@ cn_maint_task(struct work_struct *work)
 }
 
 struct cndb_cn_ctx {
-    struct cn  *cn;
-    struct map *nodemap;
+    struct cn_tree *tree;
+    struct map     *nodemap;
 };
 
-/* This function is the cndb_cn_instantiate() callback.  It is called
- * once for each kvset found in the cndb for the cn specified in the
- * callback context.
+static merr_t
+cndb_cn_ctx_init(struct cndb_cn_ctx *ctx, struct cn_tree *tree, struct cn_tree_node *root)
+{
+    struct map *nodemap;
+
+    INVARIANT(ctx);
+    INVARIANT(tree);
+    INVARIANT(root);
+
+    nodemap = map_create(CN_FANOUT_MAX);
+    if (!nodemap)
+        return merr(ENOMEM);
+
+    ctx->nodemap = nodemap;
+    ctx->tree = tree;
+
+    /* Pre-populate node map with root node, which always has ID 0. */
+    map_insert_ptr(ctx->nodemap, 0, root);
```

Comment:
map_insert_ptr() can fail?

Reply:
I'm liking the idea more and more to have a route map iterator that calls the caller's comparator to find nodes of interest...

Reply:
Memory allocation can cause failure.

Reply:
return map_insert_ptr(ctx->nodemap, 0, root);